### PR TITLE
Updating to assigna custom role based on variable instead of default-user-role

### PR DESCRIPTION
**Pull request by nitishagar**
Ref: https://github.com/balena-io/balena-api/pull/2913

Change-Type: patch

### CODE REVIEW

**Review by dfunckt — CHANGES_REQUESTED** [review]
There's no point declaring a role here that is not used by the core API -- unless there's use for it, it should be defined in the cloud API.

File: src/features/auth/hooks/assign-user-role.ts [hunk 1]
```diff
@@ -14,7 +15,7 @@ hooks.addPureHook('POST', 'resin', 'user', {
 				req: permissions.root,
 			},
 			id: {
-				name: 'default-user',
+				name: getNewUserRoleName() || 'default-user',
```
Comment (dfunckt) [inline src/features/auth/hooks/assign-user-role.ts:18]: It's impossible for `getNewUserRoleName()` to return a null-ish value, so:

```suggestion
				name: getNewUserRoleName(),
```

File: src/infra/auth/auth.ts [hunk 2]
```diff
@@ -22,6 +23,17 @@ export const userHasPermission = (
...
+let getNewUserRole: ((user: AnyObject) => string) | undefined;
+
+export const setRegistrationRoleFunc = (
+	registerFunc: (user: AnyObject) => string,
+) => {
+	getNewUserRole = registerFunc;
+};
+
+export const getRegistrationRole = (user: AnyObject) =>
+	getNewUserRole && getNewUserRole(user);
```
Comment (thgreasi) [inline src/infra/auth/auth.ts:35]:
```suggestion
export let getNewUserRole: ((user: AnyObject) => string) = () => 'default-user';

export const setRegistrationRoleFunc = (
	registerFunc: (user: AnyObject) => string,
) => {
	getNewUserRole = registerFunc;
};
```

**Review by thgreasi — COMMENTED** [review]
And with that ^ and Akis' comments fixed, LGTM.
Not convinced I like the new name of `assign-default-user-role.ts`, since this is still the code that assigns the default registration role, regardless of what the role provider returns.

File: src/infra/auth/auth.ts [hunk 3]
```diff
@@ -22,6 +23,16 @@ export const userHasPermission = (
...
 
+let getNewUserRole: (user: AnyObject) => string = () => 'default-user';
+
+export const setRegistrationRoleFunc = (
+	registerFunc: (user: AnyObject) => string,
+) => {
+	getNewUserRole = registerFunc;
+};
+
+export const getRegistrationRole = (user: AnyObject) => getNewUserRole(user);
```
Comment (thgreasi) [inline src/infra/auth/auth.ts:34]: I guess that's just so that no-one else can modify `getNewUserRole` right?

**Review by thgreasi — COMMENTED** [review]
LGTM

File: src/index.ts [hunk 4]
```diff
@@ -272,6 +273,8 @@ export interface SetupOptions {
 	onLogin?: (
 		user: Pick<DbUser, typeof defaultFindUser$select[number]>,
 	) => PromiseLike<void> | void;
+
+	getNewUserRole?: (user: AnyObject) => string;
```
Comment (dfunckt) [inline src/index.ts:277]: As per `GetUrlFunction` and `SetupFunction` above, let's give the callback a name that can be imported by clients:

```ts
  // src/infra/auth/auth.ts
  export type GetNewUserRoleFunction = (user: AnyObject) => string;
  
  // this file:
  getNewUserRole?: GetNewUserRoleFunction;
```

**Review by dfunckt — COMMENTED** [review]
Great stuff! My comment is optional, feel free to ignore.

File: src/infra/auth/auth.ts [hunk 5]
```diff
@@ -8,6 +8,7 @@ import { User } from './jwt-passport';
 
 import { getIP } from '../../lib/utils';
 import type { User as DbUser } from '../../balena-model';
+import { AnyObject } from 'pinejs-client-core';
```
Comment (Page-) [inline src/infra/auth/auth.ts:11]: This is unnecessary - it's globally available
```suggestion
```

File: src/infra/auth/auth.ts [hunk 6]
```diff
@@ -22,6 +23,16 @@ export const userHasPermission = (
 	return user.permissions.includes(permission);
 };
 
+let getNewUserRole: (user: AnyObject) => string = () => 'default-user';
+
+export const setRegistrationRoleFunc = (
+	registerFunc: (user: AnyObject) => string,
```
Comment (Page-) [inline src/infra/auth/auth.ts:29]:
```suggestion
	registerFunc: typeof getNewUserRole,
```

File: src/infra/auth/auth.ts (hunk 3 above)
Comment (Page-) [inline src/infra/auth/auth.ts:34]: I think you can inherit types like:
```suggestion
export const getRegistrationRole: typeof getNewUserRole = (user) => getNewUserRole(user);
```

**Review by dfunckt — APPROVED** [review]

**Pull request merged**